### PR TITLE
feat(wandb): support distributed modes

### LINKED ISSUE #11602: Improve multiprocessing compatibility with WandbLogger
## 🚀 Feature

Weights and Biases is updating their logging backend to better support logging in multiple processes. This issue proposes a change to the WandbLogger implementation for an optimal integration.

### Motivation

After #10896, the logger experiment consistently gets created at the first call to `logger.experiment` directly in the spawned worker processes. This is generally required because run objects for various third-party loggers are not pickleable. Secondly, depending on the logger backend, starting an experiment in the main process and logging in the worker processes could potentially create unwanted duplicate experiments/runs. However, Wandb has a special worker process running specifically just for logging. This service process can communicate with the other worker processes, **but only if it gets created in the main process** (this is a new feature being developed), which is no longer the case after #10896. Therefore we seek a solution to initiate that experiment on the main process instead, while still avoiding the pickle issue. 

### Pitch

1. Call wandb.init in the constructor of WandbLogger. 
2. Wrap it with the [rank_zero_only](https://github.com/PyTorchLightning/pytorch-lightning/blob/fe34bf2a653ebd50e6a3a00be829e3611f820c3c/pytorch_lightning/utilities/distributed.py#L47) and ensure rank > 0 returns a [mocked experiment](https://github.com/PyTorchLightning/pytorch-lightning/blob/fe34bf2a653ebd50e6a3a00be829e3611f820c3c/pytorch_lightning/loggers/base.py#L34)
3. Ensure that the logger is still [picklable](https://github.com/PyTorchLightning/pytorch-lightning/blob/fe34bf2a653ebd50e6a3a00be829e3611f820c3c/pytorch_lightning/loggers/wandb.py#L311-L318)

### Alternatives

A alternative we discussed was to introduce a hook for the logger that runs before fit spawns processes. Here, WandbLogger could get a chance to instantiate their experiment.

### Additional context

This will introduce a slight inconsistency with other logger implementations. However, it is always up to the specific logger to implement how they handle their experiment creation. As long as it adheres to the user facing api and experiments get created at the right time before users log anything, this change is imo acceptable.

cc @kptkin @raubitsj

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @awaelchli @morganmcg1 @AyushExel @borisdayma @scottire @manangoel99

### PR DESCRIPTION
**Pull request by borisdayma**
## What does this PR do?

Fixes #11602 

Improves support of distributed mode.

For now it also requires calling `wandb.require("service")` in training script.

### Does your PR introduce any breaking changes? If yes, please list them.

The run is created earlier than it used to be as it happens when we instantiate `WandbLogger` while it used to be created when first logging metric, hyper-parameters or logging any other supported method.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @awaelchli @kptkin @raubitsj who are also working on this PR
cc @morganmcg1 @scottire @manangoel99

### CODE REVIEW

**Comment by awaelchli** [comment]
Amazing! I tried and it works great so far. Couldn't push to your branch for some reason. I suggest to update the tests:

```diff
diff --git a/tests/loggers/test_wandb.py b/tests/loggers/test_wandb.py
index 85b20c562..d347d32d1 100644
--- a/tests/loggers/test_wandb.py
+++ b/tests/loggers/test_wandb.py
@@ -22,6 +22,7 @@ from pytorch_lightning import Trainer
 from pytorch_lightning.loggers import WandbLogger
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 from tests.helpers import BoringModel
+from tests.helpers.utils import no_warning_call
 
 
 @mock.patch("pytorch_lightning.loggers.wandb.wandb")
@@ -57,7 +58,7 @@ def test_wandb_logger_init(wandb):
     # verify default resume value
     assert logger._wandb_init["resume"] == "allow"
 
-    with pytest.warns(UserWarning, match="There is a wandb run already in progress"):
+    with no_warning_call(UserWarning, match="There is a wandb run already in progress"):
         _ = logger.experiment
 
     logger.log_metrics({"acc": 1.0}, step=3)
@@ -124,8 +125,10 @@ def test_wandb_pickle(wandb, tmpdir):
 def test_wandb_logger_dirs_creation(wandb, tmpdir):
     """Test that the logger creates the folders and files in the right place."""
     logger = WandbLogger(save_dir=str(tmpdir), offline=True)
-    assert logger.version is None
-    assert logger.name is None
+
+    # the wandb Run gets initialized on instantiation, version and name should be available
+    assert logger.version is not None
+    assert logger.name is not None
 
     # mock return values of experiment
     wandb.run = None
```

File: pytorch_lightning/loggers/wandb.py [hunk 1]
```diff
@@ -307,11 +307,16 @@ def __init__(
         self._save_dir = self._wandb_init.get("dir")
         self._name = self._wandb_init.get("name")
         self._id = self._wandb_init.get("id")
+        # start wandb run (to create an attach_id for distributed modes)
```
[the remark this answers is not in the document]
Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py:311]: Great idea! I added it.

**Review by awaelchli — APPROVED** [review]

File: tests/loggers/test_wandb.py [hunk 2]
```diff
@@ -52,14 +52,12 @@ def test_wandb_logger_init(wandb):
...
     wandb.run = wandb.init()
-    logger = WandbLogger()
+    with pytest.warns(UserWarning, match="There is a wandb run already in progress"):
+        logger = WandbLogger()
 
     # verify default resume value
     assert logger._wandb_init["resume"] == "allow"
 
-    with pytest.warns(UserWarning, match="There is a wandb run already in progress"):
-        _ = logger.experiment
```
[the remark this answers is not in the document]
Comment (borisdayma) [reply tests/loggers/test_wandb.py:61]: Not sure I understood:
* if there is no existing `wandb.run`, a new experiment is created (no warning), this is tested at line 45.
* if there is already an existing `wandb.run`, then we get a warning saying that this is what will be used. Full warning is `There is a wandb run already in progress and newly created instances of "WandbLogger" will reuse this run. If this is not desired, call "wandb.finish()" before instantiating "WandbLogger".`

So if I test that the previous `wandb.init()` gets used correctly and picked by the logger, we should have a warning.

[the remark this answers is not in the document]
Comment (borisdayma) [reply tests/loggers/test_wandb.py:61]: Oh I may have understood!
I'm gonna add right after that calling `logger.experiment` does not create a new run.

File: pytorch_lightning/loggers/wandb.py [hunk 3]
```diff
@@ -33,6 +33,7 @@
 
 _WANDB_AVAILABLE = _module_available("wandb")
 _WANDB_GREATER_EQUAL_0_10_22 = _compare_version("wandb", operator.ge, "0.10.22")
```
Comment (justusschock) [inline pytorch_lightning/loggers/wandb.py:35]: nit: Should we add these to `utilities/imports.py` (@awaelchli) ?

**Review by justusschock — APPROVED** [review]
LGTM in general, some very minor nit :)

File: pytorch_lightning/loggers/wandb.py [hunk 4]
```diff
@@ -490,6 +505,7 @@ def _scan_and_log_checkpoints(self, checkpoint_callback: "ReferenceType[ModelChe
                             "save_top_k",
                             "save_weights_only",
                             "_every_n_train_steps",
+                            "_every_n_val_epochs",
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/wandb.py:508]: where does this come from? can't find any such attribute in model checkpoint callback.

File: pytorch_lightning/loggers/wandb.py [hunk 5]
```diff
@@ -490,6 +505,7 @@ def _scan_and_log_checkpoints(self, checkpoint_callback: "ReferenceType[ModelChe
...
                         ]
                         # ensure it does not break if `ModelCheckpoint` args change
                         if hasattr(checkpoint_callback, k)
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/wandb.py:511]: quick ques: any reason why we need to have this check here?
if something is changed in ModelCheckpoint, shouldn't the change be made here considering both of them are on master.

File: tests/loggers/test_wandb.py [hunk 6]
```diff
@@ -18,10 +18,16 @@
...
+# ensure we consider recent wandb versions
+import pytorch_lightning.loggers.wandb as pl_wandb
 from pytorch_lightning import Trainer
 from pytorch_lightning.loggers import WandbLogger
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 from tests.helpers import BoringModel
+from tests.helpers.utils import no_warning_call
+
+pl_wandb._WANDB_GREATER_EQUAL_0_10_22 = True
+pl_wandb._WANDB_GREATER_EQUAL_0_12_10 = True
```
Comment (rohitgr7) [inline tests/loggers/test_wandb.py:30]: is it possible to mock them instead for the test?

File: pytorch_lightning/loggers/wandb.py (hunk 4 above)
Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py:508]: My mistake!

File: pytorch_lightning/loggers/wandb.py (hunk 5 above)
Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py:511]: It's just an extra security, there are so many cases to consider that I  just wanted to make sure it would not break

**Comment by borisdayma** [comment]
Confused about the pre-commit error.
I have the hook installed locally and same isort version.

File: tests/loggers/test_wandb.py (hunk 6 above)
Comment (awaelchli) [reply tests/loggers/test_wandb.py:29]: yep, monkeypatch.setattr would be a better way to mock the attribute

**Comment by rohitgr7** [comment]
> Confused about the pre-commit error. I have the hook installed locally and same isort version.

pre-commit errors seem correct since there is an import that isn't used.

**Comment by borisdayma** [comment]
I was not able to understand the problem with `isort`.

Also  I tried using `monkeypatch` for setting `wandb` version or `_WANDB_GREATER_EQUAL_XXX` but ran into a bunch of errors…

File: pytorch_lightning/loggers/wandb.py [hunk 7]
```diff
@@ -27,16 +26,14 @@
...
 from pytorch_lightning.utilities.rank_zero import rank_zero_only, rank_zero_warn
 
-_WANDB_AVAILABLE = _module_available("wandb")
-_WANDB_GREATER_EQUAL_0_10_22 = _compare_version("wandb", operator.ge, "0.10.22")
-
 try:
-    import wandb
     from wandb.wandb_run import Run
+
+    import wandb
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/wandb.py:36]:
```suggestion
    import wandb
    from wandb.wandb_run import Run
```
this will solve the pre-commit.
will look into monkeypatch tomorrow. Unfortunately can't commit to your PR.

Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py]: Strange… `isort` wants it in the other order for me…

Looks like the `wandb` fork is private, probably why you can't commit there. I could PR through my own fork.

Comment (awaelchli) [reply pytorch_lightning/loggers/wandb.py]: could it be that you are running a different version of isort?

Comment (borisdayma) [reply pytorch_lightning/loggers/wandb.py]: I checked that but was the same...
Could be that wandb cannot be installed or I have some local settings messing this up.

**Comment by rohitgr7** [comment]
hey @borisdayma 

here is the updated code with monkeypatch you can use: https://pastebin.com/hvkyGxqc

**Comment by borisdayma** [comment]
Many thanks @rohitgr7 
I'll make sure to use a non-protected fork in the future!

File: pytorch_lightning/loggers/wandb.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loggers/wandb.py]: I disagree with this suggestion.

Why wouldn't the WandB file define its own available check?
Doing so keeps all wandb things together and avoids one dependency between files.
No other file should need to import this variable since the code is all condensed into this logger.

Comment (rohitgr7) [reply pytorch_lightning/loggers/wandb.py]: well, I thought we keep all the imports in one place. like JSONARGPARSE_AVAILABLE,  which is used only for cli, Neptune_Available (another loggeR), BAGUA_AVAILABLE, ... etc.

Comment (carmocca) [reply pytorch_lightning/loggers/wandb.py]: For legacy, IMO it would be cleaner to move as many as possible out

Similarly to my comment here https://github.com/PyTorchLightning/pytorch-lightning/pull/11797#discussion_r802963748

That'd included those 4 you mentioned too

**Review by carmocca — APPROVED** [review]
